libbeat/tests/compose/runner.go

```diff
@@ -0,0 +1,119 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package compose
+
+import (
+	"fmt"
+	"math/rand"
+	"os"
+	"sort"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+type TestRunner struct {
+	Service  string
+	Options  map[string][]string
+	Parallel bool
+	Timeout  int
+}
+
+type Suite map[string]func(t *testing.T, host string)
+
+func (r *TestRunner) scenarios() []map[string]string {
+	n := 1
+	for _, values := range r.Options {
+		n *= len(values)
+	}
+
+	scenarios := make([]map[string]string, n)
+	for variable, values := range r.Options {
+		v := 0
+		for i, s := range scenarios {
+			if s == nil {
+				s = make(map[string]string)
+				scenarios[i] = s
+			}
+			s[variable] = values[v]
+			v = (v + 1) % len(values)
+		}
+	}
+
+	return scenarios
+}
+
+func (r *TestRunner) Run(t *testing.T, tests Suite) {
+	timeout := r.Timeout
+	if timeout == 0 {
+		timeout = 300
+	}
+
+	for _, s := range r.scenarios() {
```

Comment:
Would be nice to still have something like `NO_COMPOSE=1` that can be set if the environment is already running on the local machine so it will not start up all docker containers. Perhaps that still exists ...

Reply:
I have just added something like this, look at 89f5f0e94705e392683921f14f2340586037dbec

I prefer to do it as a host override instead of only a feature disable for two reasons:
- Depending on how you are running the environment you want to test with you are going to need to override the host in any case
- "Running environment" is going to be trickier now, as each service can be running with multiple versions and configurations.